crates/history/src/browser.rs

```diff
@@ -0,0 +1,383 @@
+use std::borrow::Cow;
+use std::cell::RefCell;
+use std::fmt;
+use std::rc::{Rc, Weak};
+
+use gloo_events::EventListener;
+use gloo_utils::window;
+#[cfg(feature = "serialize")]
+use serde::de::DeserializeOwned;
+#[cfg(feature = "serialize")]
+use serde::Serialize;
+use wasm_bindgen::{JsValue, UnwrapThrowExt};
+
+#[cfg(feature = "serialize")]
+use crate::error::HistoryResult;
+use crate::history::History;
+use crate::listener::HistoryListener;
+use crate::location::Location;
+
+type WeakCallback = Weak<dyn Fn()>;
+
+/// A [`History`] that is implemented with [`web_sys::History`] that provides native browser
+/// history and state access.
+#[derive(Clone)]
+pub struct BrowserHistory {
+    inner: web_sys::History,
+    callbacks: Rc<RefCell<Vec<WeakCallback>>>,
+}
+
+impl fmt::Debug for BrowserHistory {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("BrowserHistory").finish()
+    }
+}
+
+impl PartialEq for BrowserHistory {
+    fn eq(&self, _rhs: &Self) -> bool {
+        // All browser histories are created equal.
+        true
+    }
+}
+
+impl History for BrowserHistory {
+    type Location = BrowserLocation;
+
+    fn len(&self) -> usize {
+        self.inner.length().expect_throw("failed to get length.") as usize
+    }
+
+    fn go(&self, delta: isize) {
+        self.inner
+            .go_with_delta(delta as i32)
+            .expect_throw("failed to call go.")
+    }
+
+    fn push<'a>(&self, route: impl Into<Cow<'a, str>>) {
+        let url = route.into();
+        self.inner
+            .push_state_with_url(&JsValue::NULL, "", Some(&url))
+            .expect("failed to push state.");
```

Comment:
```suggestion
            .expect_throw("failed to push state.");
```

Please do the same for other `expect`s too